pulumi/backuprunner/src/scaling.ts

```diff
@@ -0,0 +1,63 @@
+import k8s from '@kubernetes/client-node';
+
+export const delay = (time: number) => {
+  return new Promise((resolve) => setTimeout(resolve, time));
+};
+
+export const scaleStatefulSet = async (
+  k8sAppsClient: k8s.AppsV1Api,
+  name: string,
+  namespace: string,
+  count: number,
+  waitToFinish: boolean
+) => {
+  console.log(`Scaling StatefulSet ${namespace}.${name} to ${count} replicas`);
+
+  const { body } = await k8sAppsClient.readNamespacedStatefulSet(name, namespace);
+
+  if(typeof(body.spec?.replicas) === "string"){
+    body.spec.replicas = count
+  }else{
+    console.warn(`Invalid replica count: ${body.spec?.replicas}`)
+  }
+
+  await k8sAppsClient.replaceNamespacedStatefulSet(name, namespace, body);
+
+  if (waitToFinish) {
+    await waitForScalingToFinish(k8sAppsClient, name, namespace, count);
+  }
+};
+
+
+const waitForScalingToFinish = async (
+  k8sAppsClient: k8s.AppsV1Api,
+  name: string,
+  namespace: string,
+  count: number
+) => {
+  const max_iterations = 100;
+
+  var done = false;
+  var iterations = 0;
+  
+  while (!done) {
```

Comment:
two thoughts here:
1) we could pass `terminationGracePeriod` in and just sleep for that amount of time, check state and proceed assuming it scaled down as expected within termination period (avoid looping/retry and just know it will take 2 mins to snapshot)
2) just use a for loop to simplify/reduce lets a bit:
```ts
  for (let i = 0; i < 100; i++) {
    // termination grace period is 120s so need to account for that
    console.log(`Waiting for ${name} to shut down...`)

    await delay(3000)
    const status = await k8sAppsClient.readNamespacedStatefulSetStatus(name, namespace)

    if (status.body.status?.availableReplicas === count) {
      console.log(`Scaling finished - ${namespace}.${name} availableReplicas is now ${count}`)
      return
    }
  }

  console.log(`Scaling StatefulSet ${namespace}.${name} has timed out`)
  process.exit(1)
```

Reply:
Even though the code is dumb and ugly, I'm a fan of simplicity. There's no way that code breaks due to some weird k8s race condition 2 releases down the road. It would definitely be nicer to pass the value down, but it's slightly more complex and prone to misconfiguration. In the 2nd case we either wait far too long or run into the same issue - there's a need to tweak the grace period. 

Ultimately I'd prefer keeping it as is or alternatively pulling in some util library to handle this for us, as there's very little room for error here and it minimizes downtime. Maybe just some simply exponential backoff like https://www.npmjs.com/package/exponential-backoff?

Reply:
I can get behind the bullet proof retry loop over grace period config. What makes you think that about the second case? Should be the exact same logic as your while loop, just as a for loop to reduce some extra variable overhead.

Reply:
didn't notice the return, I see it now, all clear